src/diff/index.js

```diff
@@ -368,9 +368,11 @@ function diffElementNodes(
 			if (
 				'value' in newProps &&
 				newProps.value !== undefined &&
-				newProps.value !== dom.value
+				newProps.value !== dom.value &&
+				newProps.defaultValue !== newProps.value
```

Comment:
I think these aren't needed

Reply:
Thanks for the feedback @JoviDeCroock ! Without that extra check, the issue stills happening. The `newProps.value` is taking the value of the `defaultValue` when the originary props.value are `undefined`. Perhaps the issue is here, however I thought this is in this way on purpose so is why I decided to check `newProps.defaultValue !== newProps.value`.


I tried to do a failing test, but I failed in that... I can reproduce it manually in the demo folder, or here: https://stackblitz.com/edit/preact-bug-defaultvalue, typing something in the input text.

This is my attempt to do the failing test:

```js
it('should keep value of uncontrolled inputs using defaultValue', () => {
	// See https://github.com/preactjs/preact/issues/2391
		
	class Input extends Component {
		render() {
			return <input type="text" defaultValue="bar" onChange={() => this.forceUpdate()} />
		}
	}

	render(<Input />, scratch);
	expect(scratch.firstChild.value).to.equal('bar');
	scratch.firstChild.focus();
	scratch.firstChild.value = 'foo';
	rerender();
	expect(scratch.firstChild.value).to.equal('foo');
});
```

However this is always passing, because after  `scratch.firstChild.value = 'foo';` line is not firing the onChange...

Reply:
You are right. Is not need it :)

Reply:
> … because after scratch.firstChild.value = 'foo'; line is not firing the onChange...

@aralroca Hey, try dispatching event manually with `scratch.firstChild.dispatchEvent(new InputEvent('input'))`. I think it should work

Reply:
Perfect! Thanks @Sasha-Sorokin